### PR TITLE
Added functions for refunding deposits in Typescript

### PR DESCRIPTION
**Pull request by tomaszslabon**
This PR adds functions for creating and submitting a deposit refund transaction in Typescript.
The refund transaction has one input (the deposit UTXO to be refunded) and one output which
returns the deposited amount of Bitcoin to the depositor (minus fee).

Additionally, a script was added that can call this function.

### CODE REVIEW

File: typescript/test/deposit-refund.test.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            depositRefundOfWitnessDepositAndNonWitnessOutput.expectedRefund
+              .transactionHash
+          )
+        })
+      }
+    )
+  })
+
+  describe("assembleDepositRefundTransaction", () => {
+    // TODO
```
Comment (tomaszslabon) [inline typescript/test/deposit-refund.test.ts:164]: We decided to add unit tests for `assembleDepositRefundTransaction` in the follow-up PR.

File: typescript/src/deposit-refund.ts [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+ * @param deposit - Data of the deposit.
+ * @param refunderKeyRing - Key ring created using the refunder's private key.
+ * @returns Empty promise.
+ */
+async function signP2WSHDepositInput(
+  transaction: any,
+  inputIndex: number,
+  deposit: Deposit,
+  refunderKeyRing: any
+): Promise<void> {
```
Comment (dimpar) [inline typescript/src/deposit-refund.ts:268]: Do we need to return an empty promise here and in the "signP2SHDepositInput" function?

File: typescript/src/deposit-refund.ts [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+      utxoWithRaw,
+      deposit,
+      refunderAddress,
+      refunderPrivateKey
+    )
+
+  // Note that `broadcast` may fail silently (i.e. no error will be returned,
+  // even if the transaction is rejected by other nodes and does not enter the
+  // mempool, for example due to an UTXO being already spent).
+  await bitcoinClient.broadcast(rawTransaction)
```
Comment (dimpar) [inline typescript/src/deposit-refund.ts:66]: Can we wrap this call in the `try catch`? Or are you saying that even `try catch` won't catch an error?

File: typescript/test/deposit-refund.test.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+import { submitDepositRefundTransaction } from "../src/deposit-refund"
+import { TransactionHash, RawTransaction } from "./bitcoin"
+import {
+  refunderPrivateKey,
+  depositRefundOfWitnessDepositAndWitnessRefunderAddress,
+  depositRefundOfNonWitnessDepositAndWitnessRefunderAddress,
+  depositRefundOfWitnessDepositAndNonWitnessRefunderAddress,
+} from "./data/deposit-refund"
+
+describe("Sweep", () => {
```
Comment (dimpar) [inline typescript/test/deposit-refund.test.ts:17]: why `Sweep`?

File: typescript/scripts/refund.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    ),
+    outputIndex: 0,
+    value: BigNumber.from(1000000),
+  }
+
+  const recipientAddress = "tb1qxe4lldal9u2paj2tuessttchs9u9ds0e5sy3v4"
+
+  const refundTxHash = await submitDepositRefundTransaction(
+    client,
+    BigNumber.from(1520),
```
[the remark this answers is not in the document]
Comment (tomaszslabon) [reply typescript/scripts/refund.ts:120]: If the fee was very low, the miners would possibly ignore this transaction and wouldn't mine it.
If it was very high, BTC would be wasted unnecessary on the fee.

`1520` is just a value that makes it sure the transaction will easily enter the testnet blockchain.
On mainnet we should ideally try to estimate this number (e.g. using some estimate fee function).

File: typescript/src/deposit-refund.ts [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+/**
+ * Submits a deposit refund by creating and broadcasting a Bitcoin P2(W)PKH
+ * deposit refund transaction.
+ * @param bitcoinClient - Bitcoin client used to interact with the network.
+ * @param fee - the value that will be subtracted from the deposit UTXO being
+ *        refunded and used as the transaction fee.
+ * @param utxo - UTXO that was created during depositing that needs be refunded.
+ * @param deposit - Details of the deposit being refunded. It should contain
+ *        the same data that was used during depositing.
+ * @param refunderAddress - Address the recipient of the refunded deposit.
```
[the remark this answers is not in the document]
Comment (tomaszslabon) [reply typescript/src/deposit-refund.ts]: Done in f0a5d8ebbbd2a48eb8df09b25923e410cc1ec128.

File: typescript/src/deposit-refund.ts [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+ * @param refunderPrivateKey - Bitcoin private key of the refunder. It must
```
[the remark this answers is not in the document]
Comment (tomaszslabon) [reply typescript/src/deposit-refund.ts]: Done in 91904f6d10c84ba1657e50e20b47e5f815dd75e7.

File: typescript/src/deposit-refund.ts [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+ *        correspond to the `refundPublicKeyHash` of the deposit script.
+ * @returns The outcome is the deposit refund transaction hash.
+ * @dev This function should be called by the refunder after `refundLocktime`
```
[the remark this answers is not in the document]
Comment (tomaszslabon) [reply typescript/src/deposit-refund.ts:32]: Done in bd1d5592e40d371232325fcf361a7b6c459c313b.

File: typescript/src/deposit-refund.ts (hunk 3 above)
Comment (tomaszslabon) [reply typescript/src/deposit-refund.ts:67]: There will be no errors. The function will fail silently.

File: typescript/test/deposit-refund.test.ts (hunk 4 above)
Comment (tomaszslabon) [reply typescript/test/deposit-refund.test.ts]: Copy-paste error. Fixed in 115e4d6e5d430293d877f5e9a6146a0043449764.

File: typescript/src/deposit-refund.ts (hunk 2 above)
Comment (tomaszslabon) [reply typescript/src/deposit-refund.ts]: Removed in e44bd1a76230b07e931846f8cddd5b960a66caca.

File: typescript/scripts/refund.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (dimpar) [reply typescript/scripts/refund.ts:120]: ok, let's leave it to a user then. I'll add some description in the following PR.

**Review by dimpar — APPROVED** [review]

**Pull request merged**